### PR TITLE
Add buildStream and tests

### PR DESCRIPTION
**Pull request by brinco80**
The idea is to provide a stream for FB pagination (feed, comments, likes, etc).  Still I think it needs refactoring (probably create a new module), but it's a working version.

Also to be able to test it I needed to add an external access_token.  I someone knows a better way, please step up.

### CODE REVIEW

File: lib/facebook.ex [hunk 1]
```diff
@@ -375,6 +377,29 @@ defmodule Facebook do
     end
   end
 
+
+  @doc """
+  Build a stream resource from a paged facebook response
+
+  ## Examples
+    iex> stream = Facebook.pageFeed(:feed, "CocaColaMx", "<Your Token>", "id,name") |> Facebook.build_stream
```
Comment (mweibel) [inline lib/facebook.ex]: the name of the function is `buildStream`, in the docs you have `build_stream`

File: lib/facebook.ex [hunk 2]
```diff
@@ -375,6 +377,29 @@ defmodule Facebook do
...
+    iex> stream |> Stream.filter( name == "CocaCola") |> Stream.take(100) |> Enum.to_list
+  """
+  @spec buildStream(Map.t) :: Enumerable.t
+  def buildStream(paged_struct) do
+    Stream.resource(
+      fn -> %__MODULE__{next: paged_struct, current: :empty} end,
+      fn(feed) ->
+	case nextPage(feed) do
```
Comment (mweibel) [inline lib/facebook.ex]: indent seems off

File: lib/facebook.ex [hunk 3]
```diff
@@ -430,4 +455,43 @@ defmodule Facebook do
...
+  defp nextPage(%__MODULE__{current: :empty, next: next} = feed) do
+    %{feed | current: next, next: :empty}
+  end
+
+  defp nextPage(%__MODULE__{current: current} = feed) do
+    case getNextPagedData(current) do
+      {:json, next_obj} -> %{feed | current: next_obj}
+      {:error, _} ->
+	Process.sleep(1_000)
+	nextPage(feed)
```
Comment (mweibel) [inline lib/facebook.ex]: indent seems off

File: lib/facebook/graph.ex [hunk 4]
```diff
@@ -48,6 +48,14 @@ defmodule Facebook.Graph do
     request(:get, url, options)
   end
 
+  @doc """
+  HTTP GET using a full URL and options
+  """
+  @spec getFullURL(url, options) :: response
+  def getFullURL(url, options \\ []) do
```
Comment (mweibel) [inline lib/facebook/graph.ex]: shouldn't we instead just make `request/3` public?

**Comment by mweibel** [comment]
Thanks for this :)

> The idea is to provide a stream for FB pagination (feed, comments, likes, etc). Still I think it needs refactoring (probably create a new module), but it's a working version.

Sounds like a good idea 👍 
About the refactoring: why do you think so? Just because it "feels" like it needs refactoring? 

> Also to be able to test it I needed to add an external access_token. I someone knows a better way, please step up.

I think you needed that because of the page. Facebook for developers is a page which I think you need to join in order to be able to access it. Test users can't join a page and that's probably why the access token doesn't work. 
Doing that won't work in an automated way though, as the access token will expire at some point. 
We have another `testPageId` which works using the access token of the test user, however there aren't any posts on that test page.. I could create some posts there though, just for testing. What do you think?

(I can also work on that, but probably not before next week)

File: lib/facebook.ex (hunk 1 above)
Comment (brinco80) [inline lib/facebook.ex]: Yeah...I changed it from build_stream to buildStream to match the functions name style.

File: lib/facebook/graph.ex (hunk 4 above)
Comment (brinco80) [inline lib/facebook/graph.ex]: I thought of that, but I didn't want to expose to much of the Graph module.  But I think it's the better choice.

**Comment by brinco80** [comment]
> About the refactoring: why do you think so? Just because it "feels" like it needs refactoring?

IMHO I think that all the stream functionality should be in a new module, and only buildStream should be in Facebook.ex, because I think Facebook.ex is becoming too bloated, and the "low-level" stream handling fits together. 

> We have another testPageId which works using the access token of the test user, however there aren't any posts on that test page.. I could create some posts there though, just for testing. What do you think?

> (I can also work on that, but probably not before next week)

Cool, that would be great! Thanks!

Please also note that if there's a connection error, [it keeps trying after sleep one second](https://github.com/mweibel/facebook.ex/blob/stream/lib/facebook.ex#L467).  Although it works for me in production, I was thinking it should pass a error function handler and let the user to decide what to do (raise an exception, log the error, etc), but it could be a little too much.  What do you think?

**Comment by mweibel** [comment]
I like the new code. You're right: in it's own module makes more sense.

> Please also note that if there's a connection error, it keeps trying after sleep one second. Although it works for me in production, I was thinking it should pass a error function handler and let the user to decide what to do (raise an exception, log the error, etc), but it could be a little too much. What do you think?

I agree that this could be somewhat confusing and also it could retry forever (in theory). 
I gures we have three options:

a) an error handler like you described to let the user decide what to do
b) immediately error out by default
c) keep retry but in a slightly more advanced way: exponential backoff time and max retries, then error out

I somehow prefer option `b)`, as it lets the user decide and keeps the implementation simple. You?

**Comment by brinco80** [comment]
I opted to let the user decide what to do, but with retries.  By default, stream retries 3 times sleeping 1 second between each try.  But is pretty easy to implement a linear or exponential back-off, log errors, raise exceptions, etc.  I did it this way because if the user want to retry a request, he can only implement it outside the stream module, which defeats the purpose of the module.

I don't know how to test a connection error (maybe it can be done at hackney level) to provide a related test. If someone know how to do it, please let me know.

Let me know what you think!

**Comment by mweibel** [comment]
Looks good. I changed the docblock for the stream example as somehow I got an error compiling it. Not sure why, but it works now. 

Another thing I changed is the access token to use. I changed to the one I use everywhere else (the one of a test user of an app I specifically created for testing this library). The tests are currently failing because there aren't 150 posts yet and my `curl` for loop is blocked atm by Facebook ;-)

**Pull request merged**